src/client/Microsoft.Identity.Client/ManagedIdentity/ManagedIdentityClient.cs

```diff
@@ -19,38 +20,74 @@ internal class ManagedIdentityClient
     {
         private const string WindowsHimdsFilePath = "%Programfiles%\\AzureConnectedMachineAgent\\himds.exe";
         private const string LinuxHimdsFilePath = "/opt/azcmagent/bin/himds";
+
+        // Cache for the managed identity source
+        private static ManagedIdentitySource? s_cachedManagedIdentitySource;
+        private static readonly SemaphoreSlim s_credentialSemaphore = new(1, 1);
         private readonly AbstractManagedIdentity _identitySource;
 
-        public ManagedIdentityClient(RequestContext requestContext)
+        internal static async Task<ManagedIdentityClient> CreateAsync(RequestContext requestContext, CancellationToken cancellationToken = default)
         {
-            using (requestContext.Logger.LogMethodDuration())
+            if (requestContext == null)
             {
-                _identitySource = SelectManagedIdentitySource(requestContext);
+                throw new ArgumentNullException(nameof(requestContext), "RequestContext cannot be null.");
             }
+
+            requestContext.Logger?.Info("[ManagedIdentityClient] Creating ManagedIdentityClient.");
+
+            AbstractManagedIdentity identitySource = await SelectManagedIdentitySourceAsync(requestContext, cancellationToken).ConfigureAwait(false);
```

Comment:
Will this be called on every MSI request, even if token is in cache? It's now very expensive method so the results need to be cached.